aml/src/resource.rs

```diff
@@ -1,112 +1,279 @@
+                break;
             }
-        } else {
-            /*
-             * We're parsing a small descriptor. Byte 0 has the format:
-             *    | Bits        | Field             |
-             *    |-------------|-------------------|
-             *    | 0-2         | Length - n bytes  |
-             *    | 3-6         | Small item type   |
-             *    | 7           | 0 = small item    |
-             *
-             * The valid types are:
-             *      0x00-0x03: Reserved
-             *      0x04: IRQ Format Descriptor
-             *      0x05: DMA Format Descriptor
-             *      0x06: Start Dependent Functions Descriptor
-             *      0x07: End Dependent Functions Descriptor
-             *      0x08: IO Port Descriptor
-             *      0x09: Fixed Location IO Port Descriptor
-             *      0x0A: Fixed DMA Descriptor
-             *      0x0B-0x0D: Reserved
-             *      0x0E: Vendor Defined Descriptor
-             *      0x0F: End Tag Descriptor
-             */
-            unimplemented!()
         }
+
+        Ok(descriptors)
     } else {
         Err(AmlError::IncompatibleValueConversion)
     }
 }
 
-#[derive(Debug)]
+/// Parse a `ResourceDescriptor`. Returns `AmlError::IncompatibleValueConversion` if the passed value is not a
+/// `Buffer`.
+fn resource_descriptor(bytes: &[u8]) -> Result<(Option<Resource>, &[u8]), AmlError> {
+    /*
+    * If bit 7 of Byte 0 is set, it's a large descriptor. If not, it's a small descriptor.
+    */
+    if bytes[0].get_bit(7) {
+        /*
+        * We're parsing a large item. The descriptor type is encoded in Bits 0-6 of Byte 0. Valid types:
+        *      0x00: Reserved
+        *      0x01: 24-bit Memory Range Descriptor
+        *      0x02: Generic Register Descriptor
+        *      0x03: Reserved
+        *      0x04: Vendor-defined Descriptor
+        *      0x05: 32-bit Memory Range Descriptor
+        *      0x06: 32-bit Fixed Memory Range Descriptor
+        *      0x07: Address Space Resource Descriptor
+        *      0x08: Word Address Space Descriptor
+        *      0x09: Extended Interrupt Descriptor
+        *      0x0a: QWord Address Space Descriptor
+        *      0x0b: Extended Address Space Descriptor
+        *      0x0c: GPIO Connection Descriptor
+        *      0x0d: Pin Function Descriptor
+        *      0x0e: GenericSerialBus Connection Descriptor
+        *      0x0f: Pin Configuration Descriptor
+        *      0x10: Pin Group Descriptor
+        *      0x11: Pin Group Function Descriptor
+        *      0x12: Pin Group Configuration Descriptor
+        *      0x13-0x7f: Reserved
+        *
+        * Byte 1 contains bits 0-7 of the length, and Byte 2 contains bits 8-15 of the length. Subsequent
+        * bytes contain the actual data items.
+        */
+        let descriptor_type = bytes[0].get_bits(0..7);
+        let length = LittleEndian::read_u16(&bytes[1..=2]) as usize + 2;
```

Comment:
Being honest, it took me a little while to work out why this was working - so this is reading the length field (which does not include the "header" byte and length), and adding two here to include the length but not the header byte; to create a slice of one descriptor, we then add `1` to also add the header byte, but then why do we add the next one? Doesn't this pass one too many bytes to each descriptor parser?

As an example, imagine we have an Extended Interrupt Descriptor with a single interrupt number. The length field is `6`, so `length` ends up as `8`, so we end up passing bytes `0..10` to `extended_interrupt_descriptor`, which is one more than the 9 bytes of the actual descriptor (which would be `0..9`).

Fwiw, I think using `let (descriptor_bytes, remaining_bytes) = bytes.split_at(length);` would be cleaner, but it would also be fine to add `3` here and add a comment explaining what's going on, I think (and adjusting the use-sites to be correct, of course).

Reply:
Ooof, yes that's not right at all. Thanks for the feedback! I've cleaned up this code so it should make more sense now.